### PR TITLE
support memtable_insert_with_hint_prefix_extractor

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Support memtable_insert_with_hint_prefix_extractor to speedup insert.
@siddontang @disksing PTAL

### CODE REVIEW

**Comment by siddontang** [comment]
LGTM

File: test/test_rocksdb_options.rs [hunk 1]
```diff
@@ -72,3 +72,38 @@ fn test_enable_statistics() {
...
+                                                       prefix_len: 2,
+                                                   }))
+        .unwrap();
+    let db = DB::open(opts, path.path().to_str().unwrap()).unwrap();
+    let wopts = WriteOptions::new();
+
+    db.put_opt(b"k0-1", b"a", &wopts).unwrap();
+    db.put_opt(b"k0-2", b"b", &wopts).unwrap();
+    db.put_opt(b"k0-3", b"c", &wopts).unwrap();
+    assert_eq!(db.get(b"k0-1").unwrap().unwrap(), b"a");
```
Comment (disksing) [inline test/test_rocksdb_options.rs]: Any method to make sure the extractor works as expected?

Comment (zhangjinpeng1987) [inline test/test_rocksdb_options.rs]: We can't mock a class derive from class `MemTableRepFactory` like cpp.

**Comment by disksing** [comment]
LGTM

**Pull request merged**